xbmc/FileItem.cpp

```diff
@@ -2870,289 +2873,225 @@ void CFileItemList::Stack(bool stackFiles /* = true */)
+                 item->IsPlayList()))
+      {
+        // File stacking
+        std::string file{};
+        std::string filePath{};
+        URIUtils::Split(StringUtils::ToLower(item->GetPath()), filePath, file);
+        if (URIUtils::HasEncodedFilename(CURL(filePath)))
+          file = CURL::Decode(file);
+
+        // Test each item against each RegExp
+        for (auto& regExp : fileRegExps)
         {
-          std::string dvdPath = item->GetOpticalMediaPath();
-
-          if (!dvdPath.empty())
+          if (regExp.RegFind(file) != -1)
           {
-            // NOTE: should this be done for the CD# folders too?
-            item->m_bIsFolder = false;
-            item->SetPath(dvdPath);
-            item->SetLabel2("");
-            item->SetLabelPreformatted(true);
-            m_sortDescription.sortBy = SortByNone; /* sorting is now broken */
+            // Get components of file name
+            const std::string fileTitle = regExp.GetMatch(1);
+            const std::string fileVolume = regExp.GetMatch(2);
+            stackCandidates.emplace_back(FILE_CANDIDATE, fileTitle, fileVolume, &item - &m_items[0],
+                                         "", item->m_dwSize);
+            break;
           }
         }
       }
     }
-  }
-}
-
-void CFileItemList::StackFiles()
-{
-  // Precompile our REs
-  VECCREGEXP stackRegExps;
-  CRegExp tmpRegExp(true, CRegExp::autoUtf8);
-  const std::vector<std::string>& strStackRegExps = CServiceBroker::GetSettingsComponent()->GetAdvancedSettings()->m_videoStackRegExps;
-  std::vector<std::string>::const_iterator strRegExp = strStackRegExps.begin();
-  while (strRegExp != strStackRegExps.end())
-  {
-    if (tmpRegExp.RegComp(*strRegExp))
-    {
-      if (tmpRegExp.GetCaptureTotal() == 4)
-        stackRegExps.push_back(tmpRegExp);
-      else
-        CLog::Log(LOGERROR, "Invalid video stack RE ({}). Must have 4 captures.", *strRegExp);
-    }
-    ++strRegExp;
-  }
-
-  // now stack the files, some of which may be from the previous stack iteration
-  int i = 0;
-  while (i < Size())
-  {
-    CFileItemPtr item1 = Get(i);
-
-    // skip folders, nfo files, playlists
-    if (item1->m_bIsFolder
-      || item1->IsParentFolder()
-      || item1->IsNFO()
-      || item1->IsPlayList()
-      )
-    {
-      // increment index
-      i++;
-      continue;
-    }
 
-    int64_t               size        = 0;
-    size_t                offset      = 0;
-    std::string           stackName;
-    std::string           file1;
-    std::string           filePath;
-    std::vector<int>      stack;
-    VECCREGEXP::iterator  expr        = stackRegExps.begin();
+    // Sort stack candidates
+    std::sort(stackCandidates.begin(), stackCandidates.end(),
+              [](const auto& s, const auto& t)
+              {
+                if (std::get<0>(s) == std::get<0>(t)) // type (file/folder)
+                  if (std::get<1>(s) == std::get<1>(t)) // title
```

Comment:
Could you add `{` to make the code a bit more readable?

Reply:
Done